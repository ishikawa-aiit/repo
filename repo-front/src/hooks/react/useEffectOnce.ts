import { useEffect, useRef, useState } from 'react'

/**
 * 依存関係がゼロの時、useEffect（）の代わりに使用するカスタムフック
 *
 * @remarks
 * このカスタムフックはReact 18の仕様変更のため、作成しています（React 18では依存関係がゼロの時、2回レンダリングされる）
 * https://github.com/vercel/next.js/issues/35822
 */
export const useEffectOnce = (effect: () => void | (() => void)) => {
  const destroyFunc = useRef<void | (() => void)>()
  const effectCalled = useRef(false)
  const renderAfterCalled = useRef(false)
  const [_val, setVal] = useState<number>(0)

  if (effectCalled.current) {
    renderAfterCalled.current = true
  }

  useEffect(() => {
    if (!effectCalled.current) {
      destroyFunc.current = effect()
      effectCalled.current = true
    }

    setVal((val) => {
      return val + 1
    })

    return () => {
      if (!renderAfterCalled.current) {
        return
      }
      if (destroyFunc.current) {
        destroyFunc.current()
      }
    }
    // マウント時に1回だけ実行するため
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])
}
