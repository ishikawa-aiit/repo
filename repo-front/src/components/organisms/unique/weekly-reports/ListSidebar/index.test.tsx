import { render, screen, within } from '@testing-library/react'

import { ListSidebar } from '.'

jest.mock('next/router', () => {
  return {
    useRouter: () => {
      return {
        pathname: '/list',
        asPath: '/list#0',
      }
    },
  }
})

const testMenuItems = [
  {
    link: '0',
    name: 'testName1',
  },
  {
    link: '1',
    name: 'testName2',
  },
]

describe('ListSidebar', () => {
  it('menuItems内のオブジェクト数分、正常にレンダリングされている', () => {
    const { container } = render(<ListSidebar menuItems={testMenuItems} />)
    const expected = testMenuItems.length
    const { getAllByText } = within(screen.getByTestId('ListSidebar'))

    expect(getAllByText(/testName\d/)).toHaveLength(expected)
    expect(container).toMatchSnapshot()
  })
})

describe('ルーターから返ってくるパス名とmenuItemsのlink', () => {
  it('一致する場合、カレントページとなる', async () => {
    const { container } = render(<ListSidebar menuItems={testMenuItems} />)
    const { getByTestId } = within(screen.getByTestId('ListSidebar'))
    const { getAllByTestId } = within(screen.getByTestId('ListSidebar'))
    const listSidebarMenuItems = getAllByTestId('ListSidebarMenuItem')

    expect(getByTestId('ListSidebarCurrentMenuItem')).toHaveTextContent('testName1')

    listSidebarMenuItems.forEach((listSidebarMenuItem) => {
      expect(listSidebarMenuItem).not.toHaveTextContent('testName1')
    })
    expect(container).toMatchSnapshot()
  })
})
