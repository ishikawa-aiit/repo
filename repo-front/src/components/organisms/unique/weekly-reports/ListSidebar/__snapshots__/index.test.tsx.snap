// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListSidebar menuItems内のオブジェクト数分、正常にレンダリングされている 1`] = `
.emotion-0 {
  position: -webkit-sticky;
  position: sticky;
  top: 68px;
  height: calc(100vh - 68px);
  width: 180px;
  background: primary.50;
  overflow-y: scroll;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-1[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-2[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  background: primary.100;
}

<div>
  <div
    class="emotion-0"
    data-cy="ListSidebar"
    data-testid="ListSidebar"
  >
    <a
      aria-current="page"
      class="emotion-1"
      data-cy="ListSidebarCurrentMenuItem"
      data-testid="ListSidebarCurrentMenuItem"
    >
      testName1
    </a>
    <a
      class="emotion-2"
      data-cy="ListSidebarMenuItem"
      data-testid="ListSidebarMenuItem"
    >
      testName2
    </a>
  </div>
</div>
`;

exports[`ルーターから返ってくるパス名とmenuItemsのlink 一致する場合、カレントページとなる 1`] = `
.emotion-0 {
  position: -webkit-sticky;
  position: sticky;
  top: 68px;
  height: calc(100vh - 68px);
  width: 180px;
  background: primary.50;
  overflow-y: scroll;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-1[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  color: text.black;
  font-size: sm;
  gap: 4px;
  font-weight: bold;
  -webkit-padding-start: 6px;
  padding-inline-start: 6px;
  -webkit-padding-end: 6px;
  padding-inline-end: 6px;
  padding-top: 4px;
  padding-bottom: 4px;
  -webkit-text-decoration: none;
  text-decoration: none;
  transition-duration: .25s;
  cursor: pointer;
}

.emotion-2[aria-current=page] {
  background-color: primary.400;
  color: text.white;
}

.emotion-2:hover,
.emotion-2[data-hover] {
  background: primary.100;
}

<div>
  <div
    class="emotion-0"
    data-cy="ListSidebar"
    data-testid="ListSidebar"
  >
    <a
      aria-current="page"
      class="emotion-1"
      data-cy="ListSidebarCurrentMenuItem"
      data-testid="ListSidebarCurrentMenuItem"
    >
      testName1
    </a>
    <a
      class="emotion-2"
      data-cy="ListSidebarMenuItem"
      data-testid="ListSidebarMenuItem"
    >
      testName2
    </a>
  </div>
</div>
`;
